### PR TITLE
twister: Adjust platform selection logic for platform_allow

### PR DESCRIPTION
**Pull request by galak**
In CI we typically build for "default" platforms.  However some
testcases that set "platform_allow" may have no overlap with the default
platforms and thus thoses testcases will not get built.

Change the platform selection logic in these cases to set the platforms
to the list of allowed platforms ("platform_allow") in the testcase.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Comment by galak** [comment]
The intent here is to get PR cover for test like: `tests/bluetooth/init/bluetooth.init.test_ctlr_observer` That aren't getting built for PRs.

**Comment by galak** [comment]
The Buildkite failures aren't related to this change, this change just exposes those issues with are only getting caught by the nightly CI build.

**Comment by PerMac** [comment]
I think @galak that we already have this covered in twister by integration_platforms. If the sample/test has integration_platforms list and integration mode is on (-G), the integration_platforms list will replace the default platforms. I think your problem can be changed by replacing (or adding) the platform_allow list with integration_platforms. I think adding functionality of integration_platforms to platform_allow might make more confusion. Therefore I opt against introducing it and am marking this as "change requested".

**Review by PerMac — CHANGES_REQUESTED** [review]
I think @galak that we already have this covered in twister by integration_platforms. If the sample/test has integration_platforms list and integration mode is on (-G), the integration_platforms list will replace the default platforms. I think your problem can be solved by replacing (or adding) the platform_allow list with integration_platforms. I think adding the functionality of integration_platforms to platform_allow might make more confusion. Therefore I opt against introducing it and am marking this as "changes requested".

**Comment by galak** [comment]
> I think @galak that we already have this covered in twister by integration_platforms. If the sample/test has integration_platforms list and integration mode is on (-G), the integration_platforms list will replace the default platforms. I think your problem can be solved by replacing (or adding) the platform_allow list with integration_platforms. I think adding the functionality of integration_platforms to platform_allow might make more confusion. Therefore I opt against introducing it and am marking this as "changes requested".

It doesn't seem that integration_platforms works the way you describe.  For example, lets use the following as our testcase/example:

`./scripts/twister -G -s tests/bluetooth/init/bluetooth.init.test_ctlr_observer --save-test foo.tests`

If we run as is, we get:
`tests/bluetooth/init/bluetooth.init.test_ctlr_observer,arc,nsim_em,skipped,CONF_FILE=prj_ctlr_observer.conf,nsim,,0,0`

If we modify `tests/bluetooth/init/testcase.yaml` as follows:

```
diff --git a/tests/bluetooth/init/testcase.yaml b/tests/bluetooth/init/testcase.yaml
index d878616e1e..e59aedba23 100644
--- a/tests/bluetooth/init/testcase.yaml
+++ b/tests/bluetooth/init/testcase.yaml
@@ -105,6 +105,8 @@ tests:
     extra_args: CONF_FILE=prj_ctlr_observer.conf
     platform_allow: nrf52840dk_nrf52840 nrf52dk_nrf52832
       nrf51dk_nrf51422 rv32m1_vega_ri5cy
+    integration_platforms:
+      - nrf52840dk_nrf52840 nrf52dk_nrf52832 nrf51dk_nrf51422 rv32m1_vega_ri5cy
   bluetooth.init.test_ctlr_central:
     extra_args: CONF_FILE=prj_ctlr_central.conf
     platform_allow: nrf52840dk_nrf52840 nrf52dk_nrf52832
```

We get an empty `foo.tests`.

**Comment by galak** [comment]
@PerMac I'd also like to see what @nashif thoughts are on having to duplicate the same lists of platforms between `integration_platforms` and `platform_allow`

**Comment by PerMac** [comment]
@galak `integration_platforms` should be given as a sequence in the .yaml. If you add it this way:
```
diff --git a/tests/bluetooth/init/testcase.yaml b/tests/bluetooth/init/testcase.yaml
index d878616e1e..0404d1a172 100644
--- a/tests/bluetooth/init/testcase.yaml
+++ b/tests/bluetooth/init/testcase.yaml
@@ -105,6 +105,11 @@ tests:
     extra_args: CONF_FILE=prj_ctlr_observer.conf
     platform_allow: nrf52840dk_nrf52840 nrf52dk_nrf52832
       nrf51dk_nrf51422 rv32m1_vega_ri5cy
+    integration_platforms:
+      - nrf52840dk_nrf52840
+      - nrf52dk_nrf52832
+      - nrf51dk_nrf51422
+      - rv32m1_vega_ri5cy
   bluetooth.init.test_ctlr_central:
     extra_args: CONF_FILE=prj_ctlr_central.conf
     platform_allow: nrf52840dk_nrf52840 nrf52dk_nrf52832
```
you will get the desired output:
```
10:35 $ cat foo.tests 
test,arch,platform,status,extra_args,handler,handler_time,ram_size,rom_size
tests/bluetooth/init/bluetooth.init.test_ctlr_observer,arm,nrf52dk_nrf52832,,CONF_FILE=prj_ctlr_observer.conf,na,,0,0
tests/bluetooth/init/bluetooth.init.test_ctlr_observer,arm,nrf52840dk_nrf52840,,CONF_FILE=prj_ctlr_observer.conf,na,,0,0
tests/bluetooth/init/bluetooth.init.test_ctlr_observer,arm,nrf51dk_nrf51422,,CONF_FILE=prj_ctlr_observer.conf,na,,0,0
tests/bluetooth/init/bluetooth.init.test_ctlr_observer,riscv32,rv32m1_vega_ri5cy,,CONF_FILE=prj_ctlr_observer.conf,na,,0,0
```

**Comment by PerMac** [comment]
Let me elaborate why I think we should depreciate `platform_allow` and use it only in very special cases (I know it is still not like this in many cases):
The problem with `platform_allow` is that it works as a hard filter for all twister use cases. If this list is given, one has no chance to run the given test on any platform outside the list. This is very limiting. 
One of the use cases was that developers were adding this list to limit the CI scope. But then nothing else can be tested, even outside the CI. Often, the platforms in an allow list were added simply because those were the only platforms the developer tested. So even tests can run on other platforms, they are strictly blocked. It was one of the reasons why integration_platforms were introduced. To show scope for the CI tests, and not limit user use cases.
The functionality of `integration_platforms` was then modified to make this list replace default platforms in the integration mode, so it gives flexibility allowing to alter the CI testing scope on the test level. 
IMO seeing a platform in an `integration_platforms` list should give a user a "warranty", that the given platform is properly tested in the CI: it is always within the scope of the CI. What's more, I propose to use the arg from the PR also as a default for the CI: https://github.com/zephyrproject-rtos/zephyr/pull/33946 With this mode on all skips on integration platforms will be treated as errors. 
The best solution would be if a test scope would be defined based on what features a test depends on, and not by platform_allow list. Tests relying on platform_allow are harder to maintain. They require modification every time we want to be able to run it on any other platform outside the list.

**Comment by nashif** [comment]
> @PerMac I'd also like to see what @nashif thoughts are on having to duplicate the same lists of platforms between `integration_platforms` and `platform_allow`

integration_platforms should not just duplicate platform_allow. This completely defeats the purpose of `integration_platform`.

`platform_allow` signifies that this test is supported only on those platforms and this is valid, some tests or samples in fact can only run on a limited set of platforms.

`integration_platform` tells twister to pick one or more supported platforms that can run the test and execute this for coverage in CI or a constrained environment where you do not want to build and run redundant tests.

So in short, those are 2 different things. The -G option is not default  and used only in CI for PR validation. When you run without -G you should still be able to run as many tests on as many boards as you wish, `platform_allow` will then filter based on capabilities.

Is `platform_allow` usage perfect in Zephyr, no it is not, we need to clean this up for sure, but I do not think it should be deprecate and it for sure does not have the same meaning is integration_platforms.

**Comment by nashif** [comment]
I think this change is fine in general and the reason we are doing things this way is related to keeping the number of runs under control. The idea was if some test is not covered by default platforms, we should pick one of the allowed platforms to get coverage, maybe something broke along the way..

**Review by nashif — APPROVED** [review]

**Review by PerMac — CHANGES_REQUESTED** [review]
I found a corner case where the proposed logic fails. When you have both `platform_allow: platA` and `integration_platforms: platB` and run with `-G` neither of the platforms is tested nor any error is reported. You can see this e.g. by adding `platform_allow: qemu_x86` to the hello_world sample's yaml and running `./scripts/twister -T samples/hello_world/ -G` 
This patch seems to solve this problem:
```
diff --git a/scripts/pylib/twister/twisterlib.py b/scripts/pylib/twister/twisterlib.py
index 910c84115a..362ca05f65 100755
--- a/scripts/pylib/twister/twisterlib.py
+++ b/scripts/pylib/twister/twisterlib.py
@@ -3047,6 +3047,8 @@ class TestSuite(DisablePyTestCollectionMixin):
 
         for tc_name, tc in self.testcases.items():
 
+            integration = self.integration and tc.integration_platforms
+
             if tc.build_on_all and not platform_filter:
                 platform_scope = self.platforms
             elif tc.integration_platforms and self.integration:
@@ -3057,7 +3059,7 @@ class TestSuite(DisablePyTestCollectionMixin):
 
             # If there isn't any overlap between the platform_allow list and the platform_scope
             # we set the scope to the platform_allow list
-            if tc.platform_allow and not platform_filter:
+            if tc.platform_allow and not platform_filter and not integration:
                 a = set(platform_scope)
                 b = set(filter(lambda item: item.name in tc.platform_allow, self.platforms))
                 c = a.intersection(b)
@@ -3173,7 +3175,6 @@ class TestSuite(DisablePyTestCollectionMixin):
             if not instance_list:
                 continue
 
-            integration = self.integration and tc.integration_platforms
             # if twister was launched with no platform options at all, we
             # take all default platforms
             if default_platforms and not tc.build_on_all and not integration:

```
With it, the platform from the integration list gets skipped (not just overwritten) as not being part of allow list. In addition with #33946 it will be marked as an error in integration mode.

I hope I am not just nitpicking with this. I am just worried that testing scope selection is getting more convoluted when already developers were complaining about the lack of transparency there, so at least we can try to eliminate some potential issues.

**Review by PerMac — APPROVED** [review]

**Comment by nashif** [comment]
> I hope I am not just nitpicking with this. I am just worried that testing scope selection is getting more convoluted when already developers were complaining about the lack of transparency there

can you please be a bit specific about this? Where is this reported, what exactly is not transparent? Isnt much of this already documented, if not, the docs needs to be fixed.
